sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,757 @@
+                    cr, ncr, ar = rhs.split_c_nc_atom(to_the_left)  # extract factors of base to the left
+                    alar = qapply_Mul2([al, ar], [])
+                    if no_interaction(al, ar, alar):
+                        return Mul(lhs, rhs, evaluate = False) # return unevaluated
+                    else:
+                        return Mul(*(cl + cr + ncl + [alar] + ncr))
+                else: # lhs or rhs have exponents <= 1, so we cannot extract factors
+                    return Mul(lhs, rhs, evaluate = False) # return unevaluated
+            elif lhs.base == rhs: # catch special case rhs == lhs.base == PowHold(lhs.base,1)
+                return lhs.add_exp(1)
+            elif (lhs.exp - 1).is_positive: # (PowHold with 1 < exp) * anything
+                alrhs = qapply_Mul2([al, rhs], [])
+                if no_interaction(al, rhs, alrhs):
+                    pass # maybe further down rhs can act on lhs, e.g. if rhs==Id
+                else:
+                    return Mul(*(cl + ncl + [alrhs]))
+        elif isinstance(rhs, PowHold):
+            if lhs == rhs.base: # catch special case lhs == rhs.base = PowHold(rhs.base,1)
+                return rhs.add_exp(1)
+            elif (rhs.exp - 1).is_positive:
+                cr, ncr, ar = rhs.split_c_nc_atom(to_the_left) # extract factors of base to the left
+                lhsar = qapply_Mul2([lhs, ar], [])
+                if no_interaction(lhs, ar, lhsar):
+                    pass # maybe further down lhs can act on rhs, e.g. if lhs==Id
+                else:
+                    return Mul(*(cr + [lhsar] + ncr))
+
+        # case TensorProduct * TensorProduct: no *-Operator, ie. no __mul__ defined
+        if isinstance(lhs, TensorProduct) and isinstance(rhs, TensorProduct) and len(lhs.args) == len(rhs.args):
+            # Do the *-operator for TensorProduct*TensorProduct. If there are
+            # c-factors in the tensor factors these will be pulled up front automatically, making it Mul
+            result = TensorProduct(*[ cast(Expr, lf) * cast(Expr, rf) for (lf, rf) in zip(lhs.args, rhs.args)])
+            #return qapply_Mul2([result], []) #qapply'ies to the tensor product factors
+            return result # the tensor factors will be qapply'ed in slit_c_nc_atom_factor
+
+        # add more cases type1 * type2 go here if they require recursive calls to qapply()
+        # ...
+
+        # Now try to apply the _apply_operator and _apply_from_right_to methods, so
+        # these take precedence over the built-in methods below.
+        result = cast(Any, None)
+        try:
+            result = cast(Any, lhs)._apply_operator(rhs, **options)
+        except (NotImplementedError, AttributeError):
+            try:
+                result = cast(Any, rhs)._apply_from_right_to(lhs, **options)
+            except (NotImplementedError, AttributeError):

```

Comment:
It is better not to catch `AttributeError` which should only really be raised in the event of a bug. You could use `hasattr` instead but I think it would be better to have an explicit check for the type that ought to have the appropriate method. Presumably then you wouldn't need to use `Any` here.

Reply:
Precisely this lines are one of the few lines of code that I have taken from the original qapply, this mechanism to call lhs._apply_operator() is very original, and this was and is the heart of qapply.  The whole quantum package is built on the idea that operators and states define the ._apply_operator() and or the ._apply_from_right() attribute to define how to operate with each other.
I cannot get rid of the try/except for NotImplementedError as raise(NotImplemented) is all over the place at least in quantum. I could probably use hasattr, but that would just make the code longer and neither add to clarity nor speed.

Reply:
At some point I removed all instances of `except AttributeError` in the codebase because they were found to be hiding bugs. I expect that I missed this one because it is catching two exceptions (so I missed it when grepping). I can also see now that more cases have crept back into the codebase because contributors keep misusing exception handling. Removing `except AttributeError` exposed a number of bugs immediately because it's a really effective way of hiding bugs.

The problem with AttributeError in particular is that it can easily be caused by a typo anywhere in the codebase just by misspelling an attribute name somewhere. This means that you need to make sure that you only catch the exception around the exact attribute retrieval like:
```python
try:
    meth = obj.method
except AttributeError:
    # do something else
# Now call meth
result = meth()
```
Of course then it would be better written as
```python
meth = getattr(obj, 'method', None)
if meth is not None:
    result = meth()
else:
    # do something else
```
The reason that people don't want to write their code like this is because there is usually something else going on like
```python
try:
   result = obj.meth()
except AttributeError:
    # do something else
```
This though is precisely how *not* to catch AttributeError because now you are potentially catching it from inside the call to `obj.meth()` rather than the attribute lookup `obj.meth` itself.

Overly broad exception handling is a big problem in the codebase and it's important not to draw on precedent from the existing code to consider what exactly is good practice (especially with an effectively unmaintained part of the codebase).

As for `NotImplementedError` that is not an exception that should be raised in any context where the caller might want to catch it. I can see that it might be difficult to rework things if it is currently being used extensively though.

Reply:
Exception(AttributeError) has been removed, Exception(NotImplemented) had to remain. So fixed.